File: backend/src/models/Donation.js
const { DataTypes } = require('sequelize');
const { sequelize } = process.env.NODE_ENV === 'test' 
  ? require('../config/test-db') 
  : require('../config/db');
const User = require('./User');

const Donation = sequelize.define('Donation', {
  id: {
    type: DataTypes.INTEGER,
    primaryKey: true,
    autoIncrement: true,
  },
  donor_id: {
    type: DataTypes.INTEGER,
    allowNull: false,
    references: {
      model: User,
      key: 'id',
    },
    onUpdate: 'CASCADE',
    onDelete: 'CASCADE',
  },
  title: {
    type: DataTypes.STRING(100),
    allowNull: false,
    validate: {
      notEmpty: true,
      len: [1, 100],
    },
  },
  description: {
    type: DataTypes.TEXT,
    allowNull: true,
  },
  donation_type: {
    type: DataTypes.STRING(50),
    allowNull: false,
    validate: {
      notEmpty: true,
      isIn: [['FOOD', 'CLOTHES', 'MEDICINE', 'OTHER']],
    },
  },
  quantity: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: 1,
    },
  },
  unit: {
    type: DataTypes.STRING(20),
    allowNull: false,
    validate: {
      notEmpty: true,
    },
  },
  expiry_date: {
    type: DataTypes.DATE,
    allowNull: false,
    validate: {
      isDate: true,
      isAfterToday(value) {
        if (new Date(value) <= new Date()) {
          throw new Error('Expiry date must be in the future');
        }
      },
    },
  },
  pickup_address: {
    type: DataTypes.TEXT,
    allowNull: false,
    validate: {
      notEmpty: true,
    },
  },
  pickup_time: {
    type: DataTypes.DATE,
    allowNull: true,
  },
  status: {
    type: DataTypes.STRING(20),
    defaultValue: 'available',
    validate: {
      isIn: [['available', 'claiming', 'in_transit', 'completed', 'cancelled']],
    },
  },
  volunteer_id: {
    type: DataTypes.INTEGER,
    allowNull: true,
    references: {
      model: User,
      key: 'id',
    },
    onUpdate: 'SET NULL',
    onDelete: 'SET NULL',
  },
  organization_id: {
    type: DataTypes.INTEGER,
    allowNull: true,
    references: {
      model: User,
      key: 'id',
    },
    onUpdate: 'SET NULL',
    onDelete: 'SET NULL',
  },
  image_url: {
    type: DataTypes.STRING,
    allowNull: true,
  },
}, {
  tableName: 'donations',
  timestamps: true,
  createdAt: 'created_at',
  updatedAt: 'updated_at',
});

// Define associations
Donation.belongsTo(User, { as: 'donor', foreignKey: 'donor_id' });
Donation.belongsTo(User, { as: 'volunteer', foreignKey: 'volunteer_id' });
Donation.belongsTo(User, { as: 'organization', foreignKey: 'organization_id' });

User.hasMany(Donation, { as: 'donations', foreignKey: 'donor_id' });

module.exports = Donation;

File: frontend/lib/models/donation.dart
import 'user.dart';
import 'volunteer_request.dart';

class Donation {
  final int id;
  final int donorId;
  final String title;
  final String description;
  final String donationType;
  final int quantity;
  final String unit;
  final DateTime expiryDate;
  final String pickupAddress;
  final DateTime? pickupTime;
  final String status;
  final int? volunteerId;
  final int? organizationId;
  final String? imageUrl;
  final DateTime createdAt;
  final DateTime updatedAt;
  final User? donor;
  final User? volunteer;
  final User? organization;
  final List<VolunteerRequest>? volunteerRequests;

  Donation({
    required this.id,
    required this.donorId,
    required this.title,
    required this.description,
    required this.donationType,
    required this.quantity,
    required this.unit,
    required this.expiryDate,
    required this.pickupAddress,
    this.pickupTime,
    required this.status,
    this.volunteerId,
    this.organizationId,
    this.imageUrl,
    required this.createdAt,
    required this.updatedAt,
    this.donor,
    this.volunteer,
    this.organization,
    this.volunteerRequests,
  });

  factory Donation.fromJson(Map<String, dynamic> json) {
    final volunteerRequestsJson = json['volunteerRequests'] ?? json['volunteer_requests'] ?? [];
    final List<VolunteerRequest> volunteerRequestsList = [];
    
    if (volunteerRequestsJson is List) {
      for (final item in volunteerRequestsJson) {
        try {
          volunteerRequestsList.add(VolunteerRequest.fromJson(item));
        } catch (e) {
          // Skip invalid volunteer request entries
        }
      }
    }
    
    return Donation(
      id: json['id'] ?? 0,
      donorId: json['donor_id'] ?? 0,
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      donationType: json['donation_type'] ?? '',
      quantity: json['quantity'] ?? 0,
      unit: json['unit'] ?? '',
      expiryDate: DateTime.parse(json['expiry_date'] ?? DateTime.now().toIso8601String()),
      pickupAddress: json['pickup_address'] ?? '',
      pickupTime: json['pickup_time'] != null ? DateTime.parse(json['pickup_time']) : null,
      status: json['status'] ?? '',
      volunteerId: json['volunteer_id'],
      organizationId: json['organization_id'],
      imageUrl: json['image_url'],
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updated_at'] ?? DateTime.now().toIso8601String()),
      donor: json['donor'] != null ? User.fromJson(json['donor']) : null,
      volunteer: json['volunteer'] != null ? User.fromJson(json['volunteer']) : null,
      organization: json['organization'] != null ? User.fromJson(json['organization']) : null,
      volunteerRequests: volunteerRequestsList.isEmpty ? null : volunteerRequestsList,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'donor_id': donorId,
      'title': title,
      'description': description,
      'donation_type': donationType,
      'quantity': quantity,
      'unit': unit,
      'expiry_date': expiryDate.toIso8601String(),
      'pickup_address': pickupAddress,
      'pickup_time': pickupTime?.toIso8601String(),
      'status': status,
      'volunteer_id': volunteerId,
      'organization_id': organizationId,
      'image_url': imageUrl,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  Donation copyWith({
    int? id,
    int? donorId,
    String? title,
    String? description,
    String? donationType,
    int? quantity,
    String? unit,
    DateTime? expiryDate,
    String? pickupAddress,
    DateTime? pickupTime,
    String? status,
    int? volunteerId,
    int? organizationId,
    String? imageUrl,
    DateTime? createdAt,
    DateTime? updatedAt,
    User? donor,
    User? volunteer,
    User? organization,
    List<VolunteerRequest>? volunteerRequests,
  }) {
    return Donation(
      id: id ?? this.id,
      donorId: donorId ?? this.donorId,
      title: title ?? this.title,
      description: description ?? this.description,
      donationType: donationType ?? this.donationType,
      quantity: quantity ?? this.quantity,
      unit: unit ?? this.unit,
      expiryDate: expiryDate ?? this.expiryDate,
      pickupAddress: pickupAddress ?? this.pickupAddress,
      pickupTime: pickupTime ?? this.pickupTime,
      status: status ?? this.status,
      volunteerId: volunteerId ?? this.volunteerId,
      organizationId: organizationId ?? this.organizationId,
      imageUrl: imageUrl ?? this.imageUrl,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      donor: donor ?? this.donor,
      volunteer: volunteer ?? this.volunteer,
      organization: organization ?? this.organization,
      volunteerRequests: volunteerRequests ?? this.volunteerRequests,
    );
  }

  bool get isEditable => status == 'available' && !isExpired;
  bool get isExpired {
    // Check if status is expired or if expiry date has passed
    return status == 'expired' || expiryDate.isBefore(DateTime.now());
  }
  bool get isDelivered => status == 'completed';
  bool get isCurrent => status == 'available' && !isExpired;

  String get statusDisplay {
    // Check for client-side expiry first
    if (isExpired) {
      return 'Expired';
    }
    
    switch (status) {
      case 'available':
        return 'Available';
      case 'claiming':
        return 'Claiming';
      case 'in_transit':
        return 'In Transit';
      case 'completed':
        return 'Delivered';
      case 'cancelled':
        return 'Cancelled';
      case 'expired':
        return 'Expired';
      default:
        return status;
    }
  }

  String get typeDisplay {
    switch (donationType) {
      case 'FOOD':
        return 'Food';
      case 'CLOTHES':
        return 'Clothes';
      case 'MEDICINE':
        return 'Medicine';
      case 'OTHER':
        return 'Other';
      default:
        return donationType;
    }
  }
}
